openwisp_users/backends.py

```diff
@@ -0,0 +1,35 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+from django.core.exceptions import ValidationError
+from django.db.models import Q
+
+User = get_user_model()
+
+
+class UserAuthenticationBackend(ModelBackend):
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        users = User.objects.filter(
+            Q(email=username) | Q(phone_number=username) | Q(username=username)
+        )
+        # can not use users.first() because it orders the queryset before
+        # returning the first object which is not what we want
+        user = users[0] if users else None
+        if user and user.check_password(password) and self.user_can_authenticate(user):
+            return user
+        return None
+
+    def get_user(self, identifier):
+        """
+        returns an active user with the identifier supplied
+        """
+        try:
+            # By default, primary key is provided as identifier
+            # during login and logout
+            return User.objects.get(pk=identifier, is_active=True)
```

Comment:
I really don't understand where this is coming from.

Reply:
This is here because, during login and logout, django calls the `get_user` method and pass to it the `pk` of the user. infact the built in `get_user` inherited from the parent backend accepts a `pk` which it uses to filter for the user. I think I will create a new method rather than override the `get_user` to avoid this kind of confusion now and also in the future